### ISSUE TITLE
[Fail] [sig-network] DNS [It] should provide /etc/hosts entries for the cluster [LinuxOnly]

### ISSUE THREAD

**Issue by gjmzj** [kubernetes/kubernetes#80152]
**What happened**:
following instruction https://github.com/cncf/k8s-conformance/blob/master/instructions.md, 
with k8s v1.14.3, sonobuoy  v0.15.0
```
Summarizing 1 Failure:

[Fail] [sig-network] DNS [It] should provide /etc/hosts entries for the cluster [LinuxOnly] [Conformance]
/workspace/anago-v1.14.3-beta.0.37+5e53fd6bc17c0d/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/dns_common.go:508

Ran 204 of 3585 Specs in 5911.786 seconds
FAIL! -- 203 Passed | 1 Failed | 0 Pending | 3381 Skipped --- FAIL: TestE2E (5911.96s)
```
and i've read related issues #43264, and serveral others, but got little information to solve it.
**What you expected to happen**:
the test suite should pass.
**How to reproduce it (as minimally and precisely as possible)**:
setup an AIO k8s cluster: https://github.com/easzlab/kubeasz/blob/master/docs/setup/quickStart.md
do k8s-conformance test: https://github.com/cncf/k8s-conformance/blob/master/instructions.md
**Anything else we need to know?**:
**Environment**: local k8s setup with kvm vm hosts
- Kubernetes version (use `kubectl version`): 1.14.3
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): Ubuntu 1604
- Kernel (e.g. `uname -a`): 4.4.0-150
- Install tools: kubeasz(github.com/easzlab/kubeasz, ansible automation for "Kubernetes The Hard Way" installation)
- Network plugin and version (if this is a network-related bug): flannel v0.11.0-amd64
- Others:
sonobuoy logs
``` bash
[sig-network] DNS
  should provide /etc/hosts entries for the cluster [LinuxOnly] [Conformance]
  /workspace/anago-v1.14.3-beta.0.37+5e53fd6bc17c0d/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:692
[BeforeEach] [sig-network] DNS
  /workspace/anago-v1.14.3-beta.0.37+5e53fd6bc17c0d/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:149
STEP: Creating a kubernetes client
Jul 16 04:07:06.986: INFO: >>> kubeConfig: /tmp/kubeconfig-103299968
STEP: Building a namespace api object, basename dns
STEP: Waiting for a default service account to be provisioned in namespace
[It] should provide /etc/hosts entries for the cluster [LinuxOnly] [Conformance]
  /workspace/anago-v1.14.3-beta.0.37+5e53fd6bc17c0d/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:692
STEP: Running these commands on wheezy: for i in `seq 1 600`; do test -n "$$(getent hosts dns-querier-1.dns-test-service.dns-5363.svc.cluster.local)" && echo OK > /results/wheezy_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local;test -n "$$(getent hosts dns-querier-1)" && echo OK > /results/wheezy_hosts@dns-querier-1;podARec=$$(hostname -i| awk -F. '{print $$1"-"$$2"-"$$3"-"$$4".dns-5363.pod.cluster.local"}');check="$$(dig +notcp +noall +answer +search $${podARec} A)" && test -n "$$check" && echo OK > /results/wheezy_udp@PodARecord;check="$$(dig +tcp +noall +answer +search $${podARec} A)" && test -n "$$check" && echo OK > /results/wheezy_tcp@PodARecord;sleep 1; done

STEP: Running these commands on jessie: for i in `seq 1 600`; do test -n "$$(getent hosts dns-querier-1.dns-test-service.dns-5363.svc.cluster.local)" && echo OK > /results/jessie_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local;test -n "$$(getent hosts dns-querier-1)" && echo OK > /results/jessie_hosts@dns-querier-1;podARec=$$(hostname -i| awk -F. '{print $$1"-"$$2"-"$$3"-"$$4".dns-5363.pod.cluster.local"}');check="$$(dig +notcp +noall +answer +search $${podARec} A)" && test -n "$$check" && echo OK > /results/jessie_udp@PodARecord;check="$$(dig +tcp +noall +answer +search $${podARec} A)" && test -n "$$check" && echo OK > /results/jessie_tcp@PodARecord;sleep 1; done

STEP: creating a pod to probe /etc/hosts
STEP: submitting the pod to kubernetes
STEP: retrieving the pod
STEP: looking for the results for each expected name from probers
Jul 16 04:07:09.080: INFO: Unable to read wheezy_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local from pod dns-5363/dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4: the server could not find the requested resource (get pods dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4)
Jul 16 04:07:09.093: INFO: Unable to read jessie_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local from pod dns-5363/dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4: the server could not find the requested resource (get pods dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4)
Jul 16 04:07:09.100: INFO: Lookups using dns-5363/dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4 failed for: [wheezy_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local jessie_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local]

Jul 16 04:07:14.104: INFO: Unable to read wheezy_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local from pod dns-5363/dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4: the server could not find the requested resource (get pods dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4)
Jul 16 04:07:14.112: INFO: Unable to read jessie_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local from pod dns-5363/dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4: the server could not find the requested resource (get pods dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4)
Jul 16 04:07:14.119: INFO: Lookups using dns-5363/dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4 failed for: [wheezy_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local jessie_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local]

Jul 16 04:07:19.104: INFO: Unable to read wheezy_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local from pod dns-5363/dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4: the server could not find the requested resource (get pods dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4)
Jul 16 04:07:19.112: INFO: Unable to read jessie_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local from pod dns-5363/dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4: the server could not find the requested resource (get pods dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4)
Jul 16 04:07:19.120: INFO: Lookups using dns-5363/dns-test-2d15a929-a77f-11e9-9ae9-da7161b169e4 failed for: [wheezy_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local jessie_hosts@dns-querier-1.dns-test-service.dns-5363.svc.cluster.local]
```

**Comment by gjmzj** [comment]
```
Jul 14 04:37:35.685: INFO: Unable to read wheezy_hosts@dns-querier-1.dns-test-service.e2e-tests-dns-jv692.svc.cluster.local from pod e2e-tests-dns-jv692/dns-test-b217596d-a5ef-11e9-9641-7a3fe53be04e: the server could not find the requested resource (get pods dns-test-b217596d-a5ef-11e9-9641-7a3fe53be04e)
Jul 14 04:37:35.712: INFO: Unable to read jessie_hosts@dns-querier-1.dns-test-service.e2e-tests-dns-jv692.svc.cluster.local from pod e2e-tests-dns-jv692/dns-test-b217596d-a5ef-11e9-9641-7a3fe53be04e: the server could not find the requested resource (get pods dns-test-b217596d-a5ef-11e9-9641-7a3fe53be04e)
Jul 14 04:37:35.718: INFO: Lookups using e2e-tests-dns-jv692/dns-test-b217596d-a5ef-11e9-9641-7a3fe53be04e failed for: [wheezy_hosts@dns-querier-1.dns-test-service.e2e-tests-dns-jv692.svc.cluster.local jessie_hosts@dns-querier-1.dns-test-service.e2e-tests-dns-jv692.svc.cluster.local]
... 
Jul 14 04:37:35.719: INFO: Unexpected error occurred: timed out waiting for the condition
```
error log: the server could not find the requested resource (get pods dns-test-b217596d-a5ef-11e9-9641-7a3fe53be04e)
but i'm pretty sure that the pod is running during the test.

**Comment by gjmzj** [comment]
i found the test commands:
```
test -n "$$(getent hosts dns-querier-1.dns-test-service.dns-5363.svc.cluster.local)"
test -n "$$(getent hosts dns-querier-1)"
podARec=$$(hostname -i| awk -F. '{print $$1"-"$$2"-"$$3"-"$$4".dns-5363.pod.cluster.local"}');
check="$$(dig +notcp +noall +answer +search $${podARec} A)" && test -n "$$check"
check="$$(dig +tcp +noall +answer +search $${podARec} A)" && test -n "$$check"
```
then i do some related tests manually as describe here https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pods, and it works;
so why my dns test suits failed, did i miss sth

**Comment by gjmzj** [comment]
when i run the test with k8s v1.15.0 and sonobuoy v0.15.0, the same failure shows up.

**Comment by gjmzj** [comment]
update:
with k8s v1.15.1 and sonobuoy v0.15.0, the same failure shows up!

**Comment by prameshj** [comment]
Do you have a similar setup where this test can be tried by hand? Need to create a pod with a host entry - yaml similar to : https://kubernetes.io/docs/concepts/services-networking/add-entries-to-pod-etc-hosts-with-host-aliases/#adding-additional-entries-with-hostaliases

Then do "dig foo.local" and see if that works?

**Comment by gjmzj** [comment]
``` bash
cat > hostaliases-pod.yaml <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: hostaliases-pod
spec:
  restartPolicy: Never
  hostAliases:
  - ip: "127.0.0.1"
    hostnames:
    - "foo.local"
    - "bar.local"
  - ip: "10.1.2.3"
    hostnames:
    - "foo.remote"
    - "bar.remote"
  containers:
  - name: cat-hosts
    image: alpine:3.9
    command:
    - sleep
    args:
    - "3600"
EOF
kubectl apply -f hostaliases-pod.yaml
```
Examine a Pod IP and check:
```
kubectl get pod -o wide
NAME              READY   STATUS    RESTARTS   AGE   IP             NODE            NOMINATED NODE   READINESS GATES
hostaliases-pod   1/1     Running   0          14s   172.20.0.221   10.100.80.116   <none>           <none>

kubectl exec -it hostaliases-pod sh
/ # cat /etc/hosts
# Kubernetes-managed hosts file.
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
fe00::0	ip6-mcastprefix
fe00::1	ip6-allnodes
fe00::2	ip6-allrouters
172.20.0.221	hostaliases-pod

# Entries added by HostAliases.
127.0.0.1	foo.local	bar.local
10.1.2.3	foo.remote	bar.remote
/ # nslookup foo.local
nslookup: can't resolve '(null)': Name does not resolve

Name:      foo.local
Address 1: 127.0.0.1 localhost
```

**Comment by athenabot** [comment]
@gjmzj @prameshj
If this issue has been triaged, please comment `/remove-triage unresolved`.

If you aren't able to handle this issue, consider unassigning yourself and/or adding the `help-wanted` label.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by prameshj** [comment]
thanks for trying it out. Looks like the test creates a pod by setting the hostname and subdomain fields rather than host aliases.

So can you create a test-pod and set pod.Spec.Hostname = "test-hostname", pod.Spec.Subdomain = "test-subdomain" and then check that /etc/hosts has an entry for "test-hostname.test-subdmain.default.svc.cluster.local"?

You can also try "getent hosts test-hostname"

**Comment by gjmzj** [comment]
@prameshj thanks for your advice, here is more testing
```
cat > test.yaml <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
  labels:
    name: test-pod 
spec:
  hostname: test-hostname
  subdomain: test-subdomain
  containers:
  - image: ubuntu:16.04
    command:
      - sleep
      - "3600"
    name: test-pod
EOF
kubectl apply -f test.yaml
```
and check by
```
root@h06v80117:~# kubectl exec -t test-pod cat /etc/hosts
# Kubernetes-managed hosts file.
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
fe00::0	ip6-mcastprefix
fe00::1	ip6-allnodes
fe00::2	ip6-allrouters
172.20.0.220	test-hostname.test-subdomain.default.svc.cluster.local.	test-hostname
root@h06v80117:~# kubectl exec -t test-pod getent hosts test-hostname
172.20.0.220    test-hostname.test-subdomain.default.svc.cluster.local. test-hostname
```

**Comment by prameshj** [comment]
interesting, does the "getent hosts test-hostname" command work too?
If it does, then I am not sure why the test is failing.

Can you try running just the one test by specifying --ginkgo.Focus="should provide \/etc\/hosts entries for the cluster"

You can also add a sleep inside the test so that you can inspect the setup to make sure the host entries are there.

**Comment by prameshj** [comment]
Any update here?

**Comment by gjmzj** [comment]
none,
i have little time to deal with it, maybe come back when a new k8s major release becomes available.

**Comment by paralika** [comment]
Is there any solution on this issue?

**Comment by fdegir** [comment]
I'm aware that this issue is closed with no additional comments from maintainers but I just wanted to comment here in case if others end up reading the discussion.

We noticed exact same problem in our CI which uses K8s conformance tests to verify the deployment. Cluster consists of 3 nodes - 1 jumphost which is a VM and where the tests are driven. The remaining 2 nodes are baremetal nodes with with Ubuntu 1804 with 56CPUs and 128G of RAM. One of these nodes is master and the other is worker. Kubernetes version is v1.17.0.

When we run conformance tests twice in row, this test case failed during one run and passed during the next one and we decided to ignore the outcome of the test after seeing the conversation here.
